drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -35,1403 +35,1532 @@
 
-	glBindFramebuffer(GL_FRAMEBUFFER, storage->frame.current_rt->copy_screen_effect.fbo);
-	glActiveTexture(GL_TEXTURE0);
-	glBindTexture(GL_TEXTURE_2D, storage->frame.current_rt->color);
+	Item::Command *const *commands = p_item->commands.ptr();
+	//int s = p_item->commands.size();
 
-	storage->shaders.copy.bind();
-	storage->shaders.copy.set_uniform(CopyShaderGLES2::COPY_SECTION, copy_section);
+	render_batches(commands, 0, current_clip, reclip, p_material);
+}
 
-	const Vector2 vertpos[4] = {
-		Vector2(-1, -1),
-		Vector2(-1, 1),
-		Vector2(1, 1),
-		Vector2(1, -1),
-	};
+void RasterizerCanvasGLES2::_canvas_item_render_commands(Item *p_item, Item *current_clip, bool &reclip, RasterizerStorageGLES2::Material *p_material) {
 
-	const Vector2 uvpos[4] = {
-		Vector2(0, 0),
-		Vector2(0, 1),
-		Vector2(1, 1),
-		Vector2(1, 0)
-	};
+	int command_count = p_item->commands.size();
+	int command_start = 0;
 
-	const int indexpos[6] = {
-		0, 1, 2,
-		2, 3, 0
-	};
+	Item::Command *const *commands = p_item->commands.ptr();
 
-	_draw_polygon(indexpos, 6, 4, vertpos, uvpos, NULL, false);
+	// while there are still more batches to fill...
+	// we may have to do this multiple times because there is a limit of 65535
+	// verts referenced in the index buffer (each potential run of this loop)
+	while (command_start < command_count) {
+		{
+			// legacy .. just create one massive batch and render everything as before
+			bdata.batches.reset();
+			Batch *batch = _batch_request_new();
+			batch->type = Batch::BT_DEFAULT;
+			batch->num_commands = command_count;
+
+			// signify to do only one while loop
+			command_start = command_count;
+		}
 
-	storage->shaders.copy.set_conditional(CopyShaderGLES2::USE_COPY_SECTION, false);
-	storage->shaders.copy.set_conditional(CopyShaderGLES2::USE_NO_ALPHA, false);
+		render_batches(commands, 0, current_clip, reclip, p_material);
 
-	glBindFramebuffer(GL_FRAMEBUFFER, storage->frame.current_rt->fbo); //back to front
-	glEnable(GL_BLEND);
+	} // while there are still more batches to fill
 }
 
-void RasterizerCanvasGLES2::_copy_texscreen(const Rect2 &p_rect) {
+void RasterizerCanvasGLES2::join_items(Item *p_item_list, int p_z, const Color &p_modulate, Light *p_light, const Transform2D &p_base_transform) {
+	bdata.items_joined.reset();
+	bdata.item_refs.reset();
 
-	state.canvas_texscreen_used = true;
+	RIState ris;
+	ris.IG_z = p_z;
+	ris.IG_modulate = p_modulate;
+	ris.IG_light = p_light;
+	ris.IG_base_transform = p_base_transform;
 
-	_copy_screen(p_rect);
+	BItemJoined *j = 0;
 
-	// back to canvas, force rebind
-	state.using_texture_rect = false;
-	state.canvas_shader.bind();
-	_bind_canvas_texture(state.current_tex, state.current_normal);
-	_set_uniforms();
+	bool batch_break = false;
+
+	while (p_item_list) {
+
+		Item *ci = p_item_list;
+
+		bool join;
+
+		if (batch_break) {
+			// always start a new batch
+			join = false;
+
+			// could be another batch break?
+			try_join_item(ci, ris, batch_break);
```

Comment:
Why are we trying to join item after already determining we have to start a new batch?

Reply:
Agree this isn't obvious to the reader, I might put in a comment to this effect. It is because try_join_item also keeps track of the current state, so we can't miss it out (even when we know not to join), otherwise the state will get out of sync.